lib/install/client/install.go

```diff
@@ -0,0 +1,138 @@
+/*
+Copyright 2019 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package client
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/gravitational/gravity/lib/constants"
+	"github.com/gravitational/gravity/lib/defaults"
+	installpb "github.com/gravitational/gravity/lib/install/proto"
+	"github.com/gravitational/gravity/lib/state"
+	"github.com/gravitational/gravity/lib/system/service"
+	"github.com/gravitational/gravity/lib/systemservice"
+
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
+)
+
+// connect creates the installer services and returns a client.
+// It performs host validation to assert whether the host can run the installer
+func (r *InstallerStrategy) connect(ctx context.Context) (installpb.AgentClient, error) {
+	r.Info("Creating and connecting to new instance.")
+	err := r.Validate()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	err = r.installSelfAsService()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer func() {
+		if err == nil {
+			return
+		}
+		r.uninstallService()
+	}()
+	var cancel context.CancelFunc
+	ctx, cancel = context.WithTimeout(ctx, r.ConnectTimeout)
+	defer cancel()
+	client, err := installpb.NewClient(ctx, r.SocketPath, r.FieldLogger)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return client, nil
+}
+
+// installSelfAsService installs a systemd unit using the current process's command line
+// and turns on service mode
+func (r *InstallerStrategy) installSelfAsService() error {
+	req := systemservice.NewServiceRequest{
+		ServiceSpec: systemservice.ServiceSpec{
+			StartCommand: strings.Join(r.Args, " "),
+			StartPreCommands: []string{
+				removeSocketFileCommand(r.SocketPath),
+			},
+			// TODO(dmitri): run as euid?
+			User:                     constants.RootUIDString,
+			RestartPreventExitStatus: strconv.Itoa(defaults.AbortedOperationExitCode),
+			// Enable automatic restart of the service
+			Restart:  "always",
+			WantedBy: "multi-user.target",
+		},
+		NoBlock: true,
+		Unmask:  true,
```

Comment:
I used this as a means to prevent service from restarting earlier in implementation. I think I can do away with this by simply removing the unit file after the operation, so will probably remove this.